### DISCUSSION TITLE
How to require one property name or another while generating a schema?

### DISCUSSION THREAD

**Discussion by ssbarnea** [pydantic/pydantic#2405] [Question] [answered]
I need to generate a schema that allows any of two properties to be specified but that will fail if both are missing.

```
# must pass
name: foo
```

```
# must pass
src: foo
```

```
# must fail due to missing name or src
{}
```

Basically both name and src are Optional[str] by themselves but one of them needs to be present to validate the mode.

How can I achieve that? Keep in mind that i MUST be able to produce a valid schema, verification this needs to happen at schema level.

**Comment by PrettyWood** [comment, marked as answer]
Hi @ssbarnea 
```py
from typing import Union

from pydantic import BaseModel, ValidationError


class M1(BaseModel):
    name: str

class M2(BaseModel):
    src: str

class M(BaseModel):
    __root__: Union[M1, M2]

M.parse_obj({'name': 'foo'})
M.parse_obj({'src': 'foo'})
try:
    M.parse_obj({})
except ValidationError as e:
    print(e)
```
Hope it helps!

**Reply by ssbarnea** [reply]
I was already contemplating the Union trick, which for my particular case should be fine as I only have two fields to switch between.

Still, the hyper scaled model (5000) described at https://github.com/samuelcolvin/pydantic/discussions/2407 will likely choke with Unions. I will find out soon.
